### DISCUSSION TITLE
ROR shift right not working correctly in a procedure *SOLVED*

### DISCUSSION THREAD

**Discussion by drexegar** [cc65/cc65#2193] [Q&A]
Working on nes project.

When using ROR in a procedure, it doesn't work properly

Outside a procedure it works fine.

Here is a code snippet:

.proc LoadAttibutes	 ; -----------------------------
	LDA #%00000001
	ROR
	ROR
	STA TESTtemp 	;stored at zero page address $0000
	RTS
.endproc ;------------------------------------

I get a result of #$C0 in the ram which is #%11000000

When i do it outside its works like it should #$40 or #%01000000

**Comment by mrdudz** [comment]
Please post an actual code snippet that reproduces the problem

**Comment by drexegar** [comment]
> Please post an actual code snippet that reproduces the problem

ok thanks, I did some thorough testing and try this in any procedure and outside and update the results

**Comment by silverdr** [comment]
But you remember that `ROR` rotates through CARRY, don't you?

**Comment by drexegar** [comment]
> But you remember that `ROR` rotates through CARRY, don't you?

ok so I Cleared the carry but i have to rotate 3 time just to move 2 spaces, is that correct?

.proc LoadAttibutes	 ; -----------------------------
	LDA #%00000001
	CLC
	ROR
	ROR
	ROR
	STA TESTtemp 	;stored at zero page address $0000
	RTS
.endproc ;--------------------------------------------

**Comment by silverdr** [comment]
What "spaces" are you talking about? And – BTW – if you have $01 it is possible to get $C0 (in or out of `.proc`) from $01 but you won't get $40 from $01 by doing two `ROR`s. So the suggestion would be to grasp some more of 6502 assembly in general first. Like you also can't have `#$C0` (or `#%11000000`) but rather just `$C0` in your memory location, etc. Trying to answer the question, without referring to "spaces" as I don't know what you mean there, if you want to get $40 from $01 by doing `ROR`s then yes, it is correct that you need to do `CLC ROR ROR ROR`. That's how the 6502 works

**Comment by drexegar** [comment]
> What "spaces" are you talking about? And – BTW – if you have $01 it is possible to get $C0 (in or out of `.proc`) from $01 but you won't get $40 from $01 by doing two `ROR`s. So the suggestion would be to grasp some more of 6502 assembly in general first. Like you also can't have `#$C0` (or `#%11000000`) but rather just `$C0` in your memory location, etc. Trying to answer the question, without referring to "spaces" as I don't know what you mean there, if you want to get $40 from $01 by doing `ROR`s then yes, it is correct that you need to do `CLC ROR ROR ROR`. That's how the 6502 works

Thanks for your help, when i said spaces I was meaning the placements in a binary format like numbers moving over when you shift to the right. It's weird when I CLC and shift to the right once nothing happens. But I could not find anything about this in nes dev or even Google, so I'll just play around with it and get use to it.

As for #$CO yeah I meant $CO I was writing it as of I was entering it as a vaule.

UPDATE: I didnt know that ROR is a 9bit rotate that threw me off. There is Definity not enough documentation out there. Now I know what it means for the carry flag to be invovled

**Reply by silverdr** [reply]
There's quite a number of good docs on 6502

https://www.masswerk.at/6502/6502_instruction_set.html

is a good reference. This one:

https://e4aws.silverdr.com/project64/c64prg/

has exhaustive description on 6502 ISA and many more. Enjoy

**Reply by drexegar** [reply]
Thanks for the links I see some I'm not aware of, but what I meant there was not enough documentation on ROR.

I was working with attributes on the nes, and there are in order as 2x2 squares instead of rows in data, and each byte held four 2 digit binary numbers for color (%00-%11).

Since I only have 32 kb and don't want to waste 25% of the space on color tables alone, I store my tiles in RAM and add color data to each metatile and use that info to generate a color table.

Problems was I did not want to start backwards to get the data but using ASL is more logical since the data would fit perfectly that way. So I did the loop to write backwards in chunks of data and got it working.
